[WFLY-15224] JTA reference doc elaborated on interposed synchronization (#14626)
wildfly/wildfly: docs/src/main/asciidoc/_developer-guide/Jakarta_Transactions_Reference.adoc

is defined by the start and end of the method and you influence the behaviour by using
https://javaee.github.io/javaee-spec/javadocs/javax/ejb/TransactionAttribute.html[`@TransactionAttribute`].

A way to rollback the transaction is to use `setRollbackOnly()` method (see <<rolling-back>>).
As well, the transaction is rolled-back when the code throws the `java.lang.RuntimeException`.

If you want to use an application exception that rolls-back the transaction then
you need to annotate the `Exception` declaration with
https://javaee.github.io/javaee-spec/javadocs/javax/ejb/ApplicationException.html#rollback--[`@ApplicationException(rollback=true)`].

NOTE: That it is a specification violation to try to combine user and
      container managed transactions, in particular calling `EJBContext.getUserTransaction()`

>       will result in a `RuntimeException`.

Reviewer: See previous comment

==== Bean managed transactions

Using the `@TransactionManagement(TransactionManagementType.BEAN)` means
the transaction will be managed manually with the use of the JTA API.

> That's with the `UserTransaction` injections and methods on it.

Reviewer: ```suggestion
In other words, methods in the interface `UserTransaction` should be used.
```

is defined by the start and end of the method and you influence the behaviour by using
https://javaee.github.io/javaee-spec/javadocs/javax/ejb/TransactionAttribute.html[`@TransactionAttribute`].

> A way to rollback the transaction is to throw `java.lang.RuntimeException`

Author: For EJB there is nothing like `rollbackOn` and using the exception - i.e., `RuntimeException` is a standard manner to control the rollback.
I changed the text in way that `setRollbackOnly` is mentioned at the first place and added the note about the annotation `@ApplicationException`.

out of the method or use `setRollbackOnly()` method (see <<rolling-back>>).

> NOTE: It's not allowed to work with `UserTransaction` within the scope

Author: Thanks for the phrasing and replaced my text with your note.

Regarding the spec note, this is a summary of different specs in looser way in a single place. Any corner cases need to be discussed with the spec anyway.

      of container managed transaction, and it's neither permitted to
      ask for the instance with
      https://javaee.github.io/javaee-spec/javadocs/javax/ejb/EJBContext.html#getUserTransaction--[`EJBContext.getUserTransaction()`].
      You receive a `RuntimeException` in such case.

==== Bean managed transactions

Using the `@TransactionManagement(TransactionManagementType.BEAN)` means
the transaction will be managed manually with the use of the JTA API.
That's with the `UserTransaction` injections and methods on it.

NOTE: You can use injection of `EJBContext` (with `@Resource` annotation)
      to access the `UserTransaction` instance.

> NOTE: If a call is made from the container managed method to bean managed method

Reviewer 2: A few points:

Are you referring to EJB calls to another EJB.
What about the other way around.
Why is this different from calling an unannotated EJB method. The reference here to "bean managed methods" seems irrelevant since the comment would apply to calling any EJB method.

The text does not aid my understanding of EJBs and transactions, on the contrary it makes me think I am missing some subtle aspects of the API, it leaves too many unanswered questions.

Reviewer 2: I am going to finish my review at this point (ie I haven't read the rest of the document) and I will put my summary on the final review.

Author: I changed the text to use term EJB. I agree it's better. I'm not sure about my phrasing but this is just a note to warn you about some corner cases. This is not about to explain all details behind.

`Why is this different from calling an unannotated EJB method.`) because it's different. Calling from EJB container managed transaction method to EJB container managed transaction method is handled in different manner than calling EJB container managed transaction method to EJB bean managed transaction method.

If you feel I haven't explained some important parts then please provide me with a text and I will be happy to add it.
Or better when this PR is merged you can open new PR where you can explain the omitted parts.

}
----

> NOTE: The `setRollbackOnly` method can be accessed from

Reviewer: ```suggestion
NOTE: The `setRollbackOnly` method can be accessed from the
```

      https://javaee.github.io/javaee-spec/javadocs/javax/transaction/TransactionManager.html#setRollbackOnly--[`TransactionManager`]
      interface as well.

==== Transaction synchronization

JTA API gives a chance to react to events of a finishing transaction.
The transaction manager invokes registered callbacks of `beforeCompletion`
and `afterCompletion`. The callbacks are defined by implementing
interface https://javaee.github.io/javaee-spec/javadocs/javax/transaction/Synchronization.html[javax.transaction.Synchronization].

The https://javaee.github.io/javaee-spec/javadocs/javax/transaction/Synchronization.html#beforeCompletion--[`beforeCompletion`]

> callback is invoked before the transaction manager commits the global transaction. Be aware that the `beforeCompletion` call back is not invoked for transaction which is rolled-back (either with `rollback()` call or when marked with `setRollbackOnly`). The `beforeCompletion` proceeds with the transaction context.

Reviewer: ```suggestion
callback is invoked before the transaction manager commits the global transaction. Be aware that the `beforeCompletion` callback is not invoked for transactions that are rolled back (either with `rollback()` call or when marked with `setRollbackOnly`). The `beforeCompletion` proceeds with the transaction context.
```

  }

  @Override
  public void afterCompletion(int status) {
    System.out.println("Transaction finished with status " + status);
  }
}
----

The transaction synchronization can be registered from `Transaction`
interface as a standard synchronization or
from `TransactionSynchronizationRegistry` as an interposed synchronization.
Any interposed synchronization will be called after all standard synchronization
callbacks of the particular type are finished
(i.e., first all standard `beforeCompletion` callbacks are executed
and then all interposed `beforeCompletion` callbacks are executed).

Using https://javaee.github.io/javaee-spec/javadocs/javax/transaction/Transaction.html[the transaction object]
to register the synchronization.

[source,java,options="nowrap"]
----
@Stateless
public class MyBean {
  @Resource(lookup = "java:/TransactionManager")
  TransactionManager tm;

  public void method() {
    // synchronization callbacks will be called when the transaction finishes,
    // that's a time when the code leaves this method
    tm.getTransaction().registerSynchronization(new MySynchronization());
  }
}
----

> When the user runs the Stateful Session Bean he can implement

Reviewer: ```suggestion
When users run the Stateful Session Bean they can implement the
```

interface https://javaee.github.io/javaee-spec/javadocs/javax/ejb/SessionSynchronization.html[javax.ejb.SessionSynchronization],

> or just to use annotations

Reviewer: ```suggestion
or just use annotations
```

https://javaee.github.io/javaee-spec/javadocs/javax/ejb/BeforeCompletion.html[`@BeforeCompletion`] and
https://javaee.github.io/javaee-spec/javadocs/javax/ejb/AfterCompletion.html[`@AfterCompletion`],

> to define the synchronization callbacks onto the bean.

Reviewer: ```suggestion
to define the synchronization callbacks within the bean.
```

> NOTE: The session synchronization interface defines three methods.
>       One of these three, the `afterBegin`, is not connected to the transaction
>       synchronization callbacks.

Reviewer: ```suggestion
NOTE: Among the methods defined in the session synchronization interface, `afterBegin` is not connected to the transaction synchronization callbacks.
```

    Thread.sleep(42000);
  }

  // called when transaction was not marked for rollback
  @BeforeCompletion
  public void beforeCompletion() {
    System.out.println("Transaction is about to be finished, expecting commit");
  }

  @AfterCompletion
  public void afterCompletion(boolean committed) {
    System.out.println("Transaction finished with the outcome "
      + (committed ? "committed" : "rolled-back"));
  }
}
----

> For CDI one can observe the https://docs.jboss.org/weld/reference/latest/en-US/html/events.html#_transactional_observers[transactional events]

Reviewer: ```suggestion
When it comes to using the CDI, it is possible to observe the https://docs.jboss.org/weld/reference/latest/en-US/html/events.html#_transactional_observers[transactional events]
```

> or he may works with `@TransactionalScope` observing the `@BeforeDestroy` and

Reviewer: ```suggestion
or  work with `@TransactionalScope` observing the `@BeforeDestroy` and
```

`@Destroyed` events. The transaction synchronization
callbacks are used underneath in both cases.

NOTE: For more information about CDI and registration of transaction
      synchronization look at
      https://jbossts.blogspot.com/2019/04/jta-and-cdi-integration.html[the article about Narayana integration of CDI events].

To register the interposed synchronization callback, the user can inject the
https://docs.oracle.com/javaee/7/api/javax/transaction/TransactionSynchronizationRegistry.html[javax.transaction.TransactionSynchronizationRegistry].
(the mandated JNDI location for the object is at `java:comp/TransactionSynchronizationRegistry`).
The instance is bound to the currently active transaction.

[source,java,options="nowrap"]
----
@Stateless
public class MyBean {
  @Resource
  TransactionSynchronizationRegistry transactionSynchronizationRegistry;

  public void method() {
    transactionSynchronizationRegistry
      .registerInterposedSynchronization(new MySynchronization());
  }
}
----

> The transaction synchronization registry adds other useful methods which are `putResource(Object key, Object value)`

Reviewer: ```suggestion
The transaction synchronization registry adds other useful methods that are `putResource(Object key, Object value)`
```

Reviewer 2: "such as" would be best
